fuel-merkle/src/common/node.rs

```diff
@@ -28,17 +30,37 @@ pub type ChildResult<T: ParentNode> = Result<T, ChildError<T::Key, T::Error>>;
 
 #[derive(Debug, Clone)]
 #[cfg_attr(feature = "std", derive(thiserror::Error))]
-pub enum ChildError<Key, E> {
-    #[cfg_attr(feature = "std", error("Child with key {0} was not found in storage"))]
+pub enum ChildError<Key, E>
+where
+    for<'a> ChildErrorKey<&'a Key>: fmt::Display,
+{
+    #[cfg_attr(feature = "std", error("Child with key {} was not found in storage", ChildErrorKey(.0)))]
     ChildNotFound(Key),
     #[cfg_attr(feature = "std", error("Node is a leaf with no children"))]
     NodeIsLeaf,
     #[cfg_attr(feature = "std", error(transparent))]
     Error(E),
 }
 
-impl<Key, E> From<E> for ChildError<Key, E> {
+impl<Key, E> From<E> for ChildError<Key, E>
+where
+    for<'a> ChildErrorKey<&'a Key>: fmt::Display,
```

Comment:
```suggestion
    ChildErrorKey<Key>: fmt::Display,
```

I guess it should work too and we can simplify it for all implementations

Reply:
If I understand correctly, you're suggesting to use an owned type `Key` rather than `&Key`. Then, we change the constraints from:
 ```
 for<'a> ChildErrorKey<&'a Key>: fmt::Display,
 ```
 to just:
 ```
 ChildErrorKey<Key>: fmt::Display,
 ```

That could work, but with the caveat that we would need adjust `ChildError` to something like this:
```
pub enum ChildError<Key, E>
where
    Key: Copy,
    ChildErrorKey<&'a Key>: fmt::Display,
{
    #[cfg_attr(feature = "std", error("Child with key {} was not found in storage", ChildErrorKey(*.0)))]
    ChildNotFound(Key),
    ...
}
```

Note the `Copy` constraint and the dereference `ChildErrorKey(*.0)`. This is because `Key` is accessed (`.0`) as a reference in the error message. We can do that, but I figured this way we avoid the extra copy.

Reply:
Yeah, I missed that `.0` is a reference=)

But it can be done in a better way via the trait. [Here](https://github.com/bvrooman/fuel-vm/pull/2) is an example.

But why do we use aliases instead of types from the `fuel-types`?
<img width="330" alt="image" src="https://user-images.githubusercontent.com/18346821/220592170-d2b5918b-406f-428d-ab95-84a649fcda3c.png">

Our types already implement `fmt::Display` and you don't need to use custom implementation.

BTW, could you next time create a branch in the `FuelLabs` org?=) In this case, it would be simpler to suggest changes=D

Reply:
Thanks for the suggestion. I like the simpler approach. 

I would like to use the `fuel-types` types, but I will do that in a separate PR. As Tommi [mentioned](https://github.com/FuelLabs/fuel-vm/pull/341#pullrequestreview-1284781805), going between different owned types (storage key <-> hash) introduces extra copies. I think I should implement this in a separate PR so we can discuss the impact more precisely.